Compilation fails with GCC 7

using Fedora 26, which has _gcc (GCC) 7.1.1 20170526_ when following the instructions for building I get the following when running make:

```bash
Scanning dependencies of target ethash
[ 28%] Building C object libethash/CMakeFiles/ethash.dir/io.c.o
[ 28%] Building C object libethash/CMakeFiles/ethash.dir/internal.c.o
/home/singingbush/cpp-ethereum/libethash/internal.c: In function ‘ethash_full_new_internal’:
/home/singingbush/cpp-ethereum/libethash/internal.c:472:6: error: this statement may fall through [-Werror=implicit-fallthrough=]
   if (ethash_io_prepare(dirname, seed_hash, &f, (size_t)full_size, true) != ETHASH_IO_MEMO_MISMATCH) {
      ^
/home/singingbush/cpp-ethereum/libethash/internal.c:477:2: note: here
  case ETHASH_IO_MEMO_MISMATCH:
  ^~~~
cc1: all warnings being treated as errors
make[2]: *** [libethash/CMakeFiles/ethash.dir/build.make:87: libethash/CMakeFiles/ethash.dir/internal.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:495: libethash/CMakeFiles/ethash.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

--- a/eth/FarmClient.h
+++ b/eth/FarmClient.h
@@ -12,7 +12,7 @@ class FarmClient : public jsonrpc::Client
     public:
         FarmClient(jsonrpc::IClientConnector &conn, jsonrpc::clientVersion_t type = jsonrpc::JSONRPC_CLIENT_V2) : jsonrpc::Client(conn, type) {}
 
-        Json::Value eth_getWork() throw (jsonrpc::JsonRpcException)
+        Json::Value eth_getWork()
         {
             Json::Value p;
             p = Json::nullValue;
@@ -22,7 +22,7 @@ class FarmClient : public jsonrpc::Client
             else
                 throw jsonrpc::JsonRpcException(jsonrpc::Errors::ERROR_CLIENT_INVALID_RESPONSE, result.toStyledString());
         }
-        bool eth_submitWork(const std::string& param1, const std::string& param2, const std::string& param3) throw (jsonrpc::JsonRpcException)
+        bool eth_submitWork(const std::string& param1, const std::string& param2, const std::string& param3)
         {
             Json::Value p;
             p.append(param1);
@@ -34,7 +34,7 @@ class FarmClient : public jsonrpc::Client
             else
                 throw jsonrpc::JsonRpcException(jsonrpc::Errors::ERROR_CLIENT_INVALID_RESPONSE, result.toStyledString());
         }
-        bool eth_submitHashrate(const std::string& param1, const std::string& param2) throw (jsonrpc::JsonRpcException)
+        bool eth_submitHashrate(const std::string& param1, const std::string& param2)
         {
             Json::Value p;
             p.append(param1);
@@ -45,7 +45,7 @@ class FarmClient : public jsonrpc::Client
             else
                 throw jsonrpc::JsonRpcException(jsonrpc::Errors::ERROR_CLIENT_INVALID_RESPONSE, result.toStyledString());
         }
-        Json::Value eth_awaitNewWork() throw (jsonrpc::JsonRpcException)
+        Json::Value eth_awaitNewWork()
         {
             Json::Value p;
             p = Json::nullValue;
@@ -55,7 +55,7 @@ class FarmClient : public jsonrpc::Client
             else
                 throw jsonrpc::JsonRpcException(jsonrpc::Errors::ERROR_CLIENT_INVALID_RESPONSE, result.toStyledString());
         }
-        bool eth_progress() throw (jsonrpc::JsonRpcException)
+        bool eth_progress()
         {
             Json::Value p;
             p = Json::nullValue;

--- a/libethash/internal.c
+++ b/libethash/internal.c
@@ -218,10 +218,10 @@ static bool ethash_hash(
 
 		for (unsigned n = 0; n != MIX_NODES; ++n) {
 			node const* dag_node;
+			node tmp_node;
 			if (full_nodes) {
 				dag_node = &full_nodes[MIX_NODES * index + n];
 			} else {
-				node tmp_node;
 				ethash_calculate_dag_item(&tmp_node, index * MIX_NODES + n, light);
 				dag_node = &tmp_node;
 			}
@@ -446,41 +446,43 @@ ethash_full_t ethash_full_new_internal(
 		return NULL;
 	}
 	ret->file_size = (size_t)full_size;
-	switch (ethash_io_prepare(dirname, seed_hash, &f, (size_t)full_size, false)) {
-	case ETHASH_IO_FAIL:
-		// ethash_io_prepare will do all ETHASH_CRITICAL() logging in fail case
+
+	enum ethash_io_rc err = ethash_io_prepare(dirname, seed_hash, &f, (size_t)full_size, false);
+	if (err == ETHASH_IO_FAIL)
 		goto fail_free_full;
-	case ETHASH_IO_MEMO_MATCH:
-		if (!ethash_mmap(ret, f)) {
-			ETHASH_CRITICAL("mmap failure()");
-			goto fail_close_file;
-		}
-#if defined(__MIC__)
-		node* tmp_nodes = _mm_malloc((size_t)full_size, 64);
-		//copy all nodes from ret->data
-		//mmapped_nodes are not aligned properly
-		uint32_t const countnodes = (uint32_t) ((size_t)ret->file_size / sizeof(node));
-		//fprintf(stderr,"ethash_full_new_internal:countnodes:%d",countnodes);
-		for (uint32_t i = 1; i != countnodes; ++i) {
-			tmp_nodes[i] = ret->data[i];
-		}
-		ret->data = tmp_nodes;
-#endif
-		return ret;
-	case ETHASH_IO_MEMO_SIZE_MISMATCH:
+
+	if (err == ETHASH_IO_MEMO_SIZE_MISMATCH) {
 		// if a DAG of same filename but unexpected size is found, silently force new file creation
 		if (ethash_io_prepare(dirname, seed_hash, &f, (size_t)full_size, true) != ETHASH_IO_MEMO_MISMATCH) {
 			ETHASH_CRITICAL("Could not recreate DAG file after finding existing DAG with unexpected size.");
 			goto fail_free_full;
 		}
-		// fallthrough to the mismatch case here, DO NOT go through match
-	case ETHASH_IO_MEMO_MISMATCH:
+		// we now need to go through the mismatch case, NOT the match case
+		err = ETHASH_IO_MEMO_MISMATCH;
+	}
+
+	if (err == ETHASH_IO_MEMO_MISMATCH || err == ETHASH_IO_MEMO_MATCH) {
 		if (!ethash_mmap(ret, f)) {
 			ETHASH_CRITICAL("mmap failure()");
 			goto fail_close_file;
 		}
-		break;
+
+		if (err == ETHASH_IO_MEMO_MATCH) {
+#if defined(__MIC__)
+			node* tmp_nodes = _mm_malloc((size_t)full_size, 64);
+			//copy all nodes from ret->data
+			//mmapped_nodes are not aligned properly
+			uint32_t const countnodes = (uint32_t) ((size_t)ret->file_size / sizeof(node));
+			//fprintf(stderr,"ethash_full_new_internal:countnodes:%d",countnodes);
+			for (uint32_t i = 1; i != countnodes; ++i) {
+				tmp_nodes[i] = ret->data[i];
+			}
+			ret->data = tmp_nodes;
+#endif
+			return ret;
+		}
 	}
+
 
 #if defined(__MIC__)
 	ret->data = _mm_malloc((size_t)full_size, 64);

